### PR TITLE
[Feature] Partial matching text search

### LINKED ISSUE #9956: ✨ Allow partial search terms for names and emails
## ✨ Feature

When using the search bar for tables of users or pool candidates, it would be helpful if we matched against names and emails after just a few characters.

## 🕵️ Details

The way the search is currently implemented, we don't match partial search terms at all. That means, even for names and emails, you have to type the complete name or email for any users to show up. It would be much more convenient if we allowed partial matches for names and emails. A minimum of 3 characters before trying to match is reasonable.

## 🎨 Design File

Not applicable

### 🧑‍🎨 Designer

@Jerryescandon 

## 🙋‍♀️ Proposed Implementation

We currently implement the user "generalSearch" with the postgres full text search, which includes name and email columns but doesn't include partial matches. We could manually parse search terms and try to stragically add `*`s, but as an easier and more limited alternative, we could match names and emails separately, even in the general search scope. I suspect the sudocode could look something like this:

```
    public static function scopeGeneralSearch(Builder $query, ?array $searchTerms): Builder
    {
        if ($searchTerms && is_array($searchTerms)) {
            
            // do full text search stuff

            $query->orWhere(function ($query) {
                return $this->scopeName($query);
            }
           $query->orWhere(function ($query) {
                return $this->scopeEmail($query);
            }
        }

        return $query;
    }
```

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] Typing the first few characters of a user's name or email in search box leads to them appearing in User and PoolCandidate tables
- [ ] Its fine if all search terms work with partial search, but not required (check that it doesn't hurt performance)
- [ ] Its okay if "partial matching" assumes you always start typing from the beginning of the word/name, ie only does wildcard matching at the end of the search term

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #9956

## 👋 Introduction

Add some sort of partial matching for the database full text search

## 🕵️ Details

Two solutions present at the moment, want input on which one is preferred 

## 🧪 Testing

1. Ideally seed more users/candidates, then head to `/en/admin/pool-candidates`
2. Experiment with searching, switch up solutions, experiment more

## 📸 Solution 1

image

image

image

Much more finicky, expect to see more error messages

image

## 📸 Solution 2

image

image

image

Less finicky, less errors but limited partial matching

image

### CODE REVIEW

**Review by tristan-orourke — COMMENTED** [review]
Sorry for the lateness, I went down a bit of a rabbit hole about database indexes and textsearch lexemes/vectors/indexes.

Short answer: I think we should go with solution style 2.

Longer answer: the functionality that improves textsearch, like normalizing words and splitting them up into lexemes, doesn't really work with proper nouns, like names and emails. For the best optimization, we should handle names and emails separately from textsearch. It might even be worth removing names and emails from the textsearch indexing. 
Then, we can do a couple things to optimize searching by name. We can add database indexes for firstname and lastname (there's already one for email), and only add the wildcard character to the end of search terms instead of the beginning, to properly take advantage of the indexes. ie
```
# this
$query->where('first_name', 'ilike', "{$value}%")
           ->orWhere('last_name', 'ilike', "{$value}%");
# instead of
$query->where('first_name', 'ilike', "%{$value}%")
           ->orWhere('last_name', 'ilike', "%{$value}%");
```
That optimization should probably be another ticket though.

**Comment by vd1992** [comment]
Okay, thanks for giving it some thought and hope you enjoyed your readings. I will leave fleshing out the optimization issue(s) to you. 
I've narrowed it to using only solution 2 and have cleaned things up, making it ready for actual review.

File: api/app/Models/User.php [hunk 1]
```diff
@@ -913,6 +913,21 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
                 // Now that we have added a column, query builder no longer will add a * to the select.  Add all possible columns manually.
                 ->addSelect(self::$selectableColumns);
 
+            // clear characters or search operators out, then create array for existing scopes by splitting into words
+            $filterToEmptySpace = ['"', '"', '-', ':', '!'];
+            $filterToSingleSpace = [' AND ', ' OR '];
```
Comment (tristan-orourke) [inline api/app/Models/User.php:918]: combined search terms may also include "&"

File: api/app/Models/User.php [hunk 2]
```diff
@@ -913,6 +913,21 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
...
+            $filtered = str_ireplace($filterToEmptySpace, '', $combinedSearchTerm);
+            $filtered = str_ireplace($filterToSingleSpace, ' ', $filtered);
+            $arrayed = explode(' ', $filtered);
```
Comment (tristan-orourke) [inline api/app/Models/User.php:921]: Why did you split this into an array? The name scope can already handle multiple words.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Okay, I wrote a number of further test cases, to see what this is capable of. Now, I'm not saying we 100% need to support all these cases. A lot of them are about trying combine partial search with websearch operator functionality. But take a look, and let me know which of these you think its reasonable to support.

File: api/app/Models/User.php (hunk 1 above)
Comment (vd1992) [reply api/app/Models/User.php]: Aye aye https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/c1db13c2e4e816bae0faae1fb61824ffdf938dac

File: api/app/Models/User.php (hunk 2 above)
Comment (vd1992) [reply api/app/Models/User.php]: Huh, it turns out it was totally unneeded 😅 
Got rid of it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/5a2a5d68e66e889f563d8fea63f094141da38b2e

**Comment by tristan-orourke** [comment]
Sigh... yeah I missed pasting this into my comment, sorry 😩 

```
     // Test user can be partial matched by name or email by scopes added into generalSearch function
    public function testUserSearchPartialMatchingNameEmail()
    {
        $user1 = User::factory()->create([
            'first_name' => 'john',
            'last_name' => 'smith',
            'email' => 'john@test.com',
            'current_city' => 'abc',
        ]);
        $user2 = User::factory()->create([
            'first_name' => 'bob',
            'last_name' => 'johnson',
            'email' => 'bob@test.com',
            'current_city' => 'efg',
        ]);
        $user3 = User::factory()->create([
            'first_name' => 'jones',
            'last_name' => 'hall',
            'email' => 'johnson@test.com',
            'current_city' => 'hij',
        ]);

        // "john" matches three users due to names and emails
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['john'],
            ],
        ])->assertJsonCount(
            3, 'data.usersPaginated.data'
        )->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ]);

        // partial name search is case insensitive
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['JOHN'],
            ],
        ])->assertJsonCount(
            3, 'data.usersPaginated.data'
        )->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ]);
    }


    public function testUserSearchPartialNamesEmailsWithOrLogic()
    {
        $user1 = User::factory()->create([
            'first_name' => 'john',
            'last_name' => 'smith',
            'email' => 'john@test.com',
            'current_city' => 'abc',
        ]);
        $user2 = User::factory()->create([
            'first_name' => 'bob',
            'last_name' => 'johnson',
            'email' => 'bob@test.com',
            'current_city' => 'efg',
        ]);
        $user3 = User::factory()->create([
            'first_name' => 'jones',
            'last_name' => 'hall',
            'email' => 'johnson@test.com',
            'current_city' => 'hij',
        ]);

        // city "abc" OR partial name "hal" matches two users
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['abc OR hal'],
            ],
        ])->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ])->assertJsonCount(2, 'data.usersPaginated.data');

         // city "abc" OR partial email "johnson@" matches two users
         $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['abc OR johnson@test'],
            ],
        ])->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ])->assertJsonCount(2, 'data.usersPaginated.data');

        // city "abc" AND partial name "joh" matches only one user
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['abc and joh'],
            ],
        ])->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonCount(1, 'data.usersPaginated.data');

        // OR operator works for partial names
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
                'where' => [
                    'generalSearch' => ['johnso or hal'],
                ],
            ])->assertJsonFragment([
                'id' => $user2->id,
            ])->assertJsonFragment([
                'id' => $user3->id,
            ])->assertJsonCount(2, 'data.usersPaginated.data');

        // OR operator works for partial emails
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['BOB@ OR johnson@'],
            ],
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ])->assertJsonCount(2, 'data.usersPaginated.data');
    }

    public function testUserSearchPartialNamesEmailsWithAndLogic()
    {
        $user1 = User::factory()->create([
            'first_name' => 'john',
            'last_name' => 'smith',
            'email' => 'john@test.com',
            'current_city' => 'abc',
        ]);
        $user2 = User::factory()->create([
            'first_name' => 'bob',
            'last_name' => 'johnson',
            'email' => 'bob@test.com',
            'current_city' => 'efg',
        ]);
        $user3 = User::factory()->create([
            'first_name' => 'jones',
            'last_name' => 'hall',
            'email' => 'johnson@test.com',
            'current_city' => 'hij',
        ]);

        // city "abc" AND partial name "joh" matches only one user
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['abc and joh'],
            ],
        ])->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonCount(1, 'data.usersPaginated.data');
    }

    public function testUserSearchPartialNamesEmailsWithQuotes()
    {
        $user1 = User::factory()->create([
            'first_name' => 'john',
            'last_name' => 'smith',
            'email' => 'john@test.com',
            'current_city' => 'abc',
        ]);
        $user2 = User::factory()->create([
            'first_name' => 'bob',
            'last_name' => 'johnson',
            'email' => 'bob@test.com',
            'current_city' => 'efg',
        ]);
        $user3 = User::factory()->create([
            'first_name' => 'jones',
            'last_name' => 'hall',
            'email' => 'johnson@test.com',
            'current_city' => 'hij',
        ]);

        // A full name surrounded by quotes should only match one person
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['"john smith"'],
            ],
        ])->assertJsonFragment([
            'id' => $user1->id,
        ])->assertJsonCount(1, 'data.usersPaginated.data');

        // A partial email surrounded by quotes continues to match
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['"bob@test"'],
            ],
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonCount(1, 'data.usersPaginated.data');

        // negative search term works for names
        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['John -Smith'],
            ],
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ])->assertJsonCount(2, 'data.usersPaginated.data');
    }

    public function testUserSearchPartialNamesEmailsWithNegativeSearch()
    {
        $user1 = User::factory()->create([
            'first_name' => 'john',
            'last_name' => 'smith',
            'email' => 'john@test.com',
            'current_city' => 'abc',
        ]);
        $user2 = User::factory()->create([
            'first_name' => 'bob',
            'last_name' => 'johnson',
            'email' => 'bob@test.com',
            'current_city' => 'efg',
        ]);
        $user3 = User::factory()->create([
            'first_name' => 'jones',
            'last_name' => 'hall',
            'email' => 'johnson@test.com',
            'current_city' => 'hij',
        ]);

        $this->actingAs($this->platformAdmin, 'api')->graphQL(
            /** @lang GraphQL */
            '
            query getUsersPaginated($where: UserFilterInput) {
                usersPaginated(where: $where) {
                    data {
                        id
                    }
                }
            }
        ', [
            'where' => [
                'generalSearch' => ['john -smit'],
            ],
        ])->assertJsonFragment([
            'id' => $user2->id,
        ])->assertJsonFragment([
            'id' => $user3->id,
        ])->assertJsonCount(2, 'data.usersPaginated.data');
    }
```

File: api/app/Models/User.php (hunk 2 above)
Comment (vd1992) [reply api/app/Models/User.php]: Or maybe it did help...
😅 🤷

**Comment by vd1992** [comment]
Okay... Butted my head for a while with this and made some progress. It is tricky as you need all parts working in tandem. 
For example, I am finding it hard to imagine how the partial match can complement the text search for AND matching.
"abc AND johns" or "johns AND abc"
Either way needs to basically disregard city for searching and execute `whereAny([columns...], 'ilike', 'johns'`) but that means the partial match needs to infer that it can ignore 'abc'. Can't quite think of how to make that work. 

I'll need your help to improve it further. 
For now, I got some progress with OR and NOT

image

File: api/app/Models/User.php [hunk 3]
```diff
@@ -930,6 +943,35 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
...
+                $query->orWhere(function ($query) use ($value, $matchesWithoutOperator) {
+                    $query->whereAny([
+                        'first_name',
+                        'last_name',
+                        'email',
+                    ], 'ilike', "%{$value}%"
+                    );
+                    $query->where(function ($query) use ($matchesWithoutOperator) {
+                        self::scopeNegationNameAndEmail($query, $matchesWithoutOperator);
+                    });
```
Comment (vd1992) [inline api/app/Models/User.php:972]: Redundant, but I needed the scope respected on both sides of the OR

File: api/app/Models/User.php [hunk 4]
```diff
@@ -930,6 +943,35 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
...
+            // clear characters or search operators out, then array split for easy OR matching
+            $filterToEmptySpace = ['"', '"', '-', ':', '!'];
+            $filterToSingleSpace = [' AND ', ' OR ', ' & '];
+            $filtered = str_ireplace($filterToEmptySpace, '', $negationRemovedSearchTerm);
+            $filtered = str_ireplace($filterToSingleSpace, ' ', $filtered);
+            $whiteSpacingRemoved = trim($filtered);
+            $arrayed = explode(' ', $whiteSpacingRemoved);
+
+            foreach ($arrayed as $index => $value) {
+                $query->orWhere(function ($query) use ($value, $matchesWithoutOperator) {
```
Comment (vd1992) [inline api/app/Models/User.php:963]: For now, effectively makes OR the base. 
I kind of like it...?

File: api/app/Models/User.php (hunk 2 above)
Comment (tristan-orourke) [reply api/app/Models/User.php]: Yeah, now I can see it helps with the expanded test cases

File: api/app/Models/User.php [hunk 5]
```diff
@@ -930,6 +943,35 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
                 // Now that we have added a column, query builder no longer will add a * to the select.  Add all possible columns manually.
                 ->addSelect(self::$selectableColumns);
 
+            // negation setup
+            preg_match('/\-[a-zA-Z]+\b/', $combinedSearchTerm, $negationMatches);
```
Comment (tristan-orourke) [inline api/app/Models/User.php:947]: This regex can be improved. First of all, `!` can also indicate negation. Second, we don't want to ignore all non-letter characters in search terms (eg hyphens or apostrophes in names, or `@`, underscore, or dot in emails).

Try these inputs in your favourite regex tester (I use https://www.regexplanet.com/advanced/javascript/index.html)
```
john -smith
john !smith
john -bob@test.com
blast-from-the-past
-john smith
```
I think this regex will work: `(^|\s)[-!][^\s]+\b`

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Great work!

I've been thinking about it more, and I think maybe negative search terms should only work with full names/emails, not partial. Otherwise, it might be too easy to accidentally exclude names which contain a word you were trying to exclude. And that allows the search to be slightly simplified.

The tests for AND logic can be removed. I don't see a way to make those work.

Handling quotes... That's certainly not impossible, but it is a bit tricky. I wonder... what if we skipped all the extra partial matching logic if the query includes any quotes? Just let the websearch handle it. When talking about how search should work, some people assumed that quoted terms would NOT work with partial matches.

File: api/app/Models/User.php [hunk 6]
```diff
@@ -910,6 +910,19 @@ public static function scopeEquity(Builder $query, ?array $equity): Builder
         return $query;
     }
 
+    public static function scopeNegationNameAndEmail(Builder $query, ?array $negationArray): Builder
+    {
+        if (isset($negationArray) && count($negationArray) > 0) {
+            foreach ($negationArray as $index => $value) {
+                $query->whereNot('first_name', 'ilike', "%{$value}%");
+                $query->whereNot('last_name', 'ilike', "%{$value}%");
+                $query->whereNot('email', 'ilike', "%{$value}%");
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/User.php]: Yeah, makes sense to not wildcard negation as that could get really confusing if it partial matched things you didn't even think about.  
Adjusted in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/e59a420ee5a7aab934b40d0edaff138d72e522f8

File: api/tests/Feature/KeywordSearchTest.php [hunk 7]
```diff
@@ -483,4 +483,348 @@ public function testUserSearchByNameAndWorkExperienceTitle()
...
+             ', [
+                'where' => [
+                    'generalSearch' => ['john -bob@test'],
+                ],
+            ])->assertJsonFragment([
+                'id' => $user1->id,
+            ])->assertJsonFragment([
+                'id' => $user3->id,
+            ])->assertJsonCount(2, 'data.usersPaginated.data');
+    }
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/tests/Feature/KeywordSearchTest.php:905]: Good idea
Add that in with https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/123636d1e2e27c639fe881feb6a78c90ff4a0c1b

File: api/app/Models/User.php (hunk 5 above)
Comment (vd1992) [reply api/app/Models/User.php]: Regex, what a wonderful world...
I'll roll with your suggestion, it seems perfect
https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/cad0c49378aabeee5baefd6478f2239eac4f7640

**Comment by vd1992** [comment]
Okay, worked on it a bit more. I removed the AND test in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/6535723fd63513683d2f20ff36e707d39647cc8b

And also, how do you feel about this for quotations? https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/b98617f4ac94830e585d84e8bae52306584a5864

🥳 

image

File: api/app/Models/User.php [hunk 8]
```diff
@@ -930,6 +943,35 @@ public static function scopeGeneralSearch(Builder $query, ?array $searchTerms):
                 // Now that we have added a column, query builder no longer will add a * to the select.  Add all possible columns manually.
                 ->addSelect(self::$selectableColumns);
 
+            // negation setup
+            preg_match('/\-[a-zA-Z]+\b/', $combinedSearchTerm, $negationMatches);
+            $matchesWithoutOperator = array_map(fn ($string) => ltrim($string, '/\-/'), $negationMatches);
+            $query->where(function ($query) use ($matchesWithoutOperator) {
+                self::scopeNegationNameAndEmail($query, $matchesWithoutOperator);
+            });
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/User.php]: I think it should be still present, no?
So something like `"john" -smith` queries for all first or last names "john" but no "smiths", as in grab "John Steven" but not "John Smith"

[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/User.php]: Oh, now I see that it is redundant https://github.com/GCTC-NTGC/gc-digital-talent/pull/9977/commits/a6613afc27c695cff788c50e52d3dee07465dac5

**Review by tristan-orourke — APPROVED** [review]
Looks great to me!

**Pull request merged**